### ISSUE TITLE
upstream: per_try_timeout kicks in when downstream buffer is full

### ISSUE THREAD

**Issue by danzh2010** [envoyproxy/envoy#25901]
Our server ran into this issue when we use multiplexed downstream protocol (HTTP/2 in our case)  and HTTP/1 upstream with an aggressive per_try_timeout.

When the responses are large and the client reads slow, the downstream connection send buffer gets filled by a few requests and goes above high watermark and stops all upstream request from reading. On this connection there are other requests whose upstream haven't started proxying the response yet, and they are read disabled and subject to the per_try_timeout in this duration.  While the downstream connection is slowly draining its send buffer, the short per_try_timeout kicks in and starts to retry or reset the whole request.

This isn't an issue if the upstream is using HTTP/2 because HTTP/2 flow control doesn't block request/response headers. And once the response headers arrives at Envoy, per_try_timeout will not take any effect. Unfortunately HTTP/3 upstream will have the same issue.

I'm working on reproducing this issue in an e2e test, meanwhile I'd like to start the discussion about the fix here.

There are two ways to change the upstream behavior, in upstream_request.cc [DownstreamWatermarkManager::onAboveWriteBufferHighWatermark](https://github.com/envoyproxy/envoy/blob/860b750b8c6fec57bdf12c58f872defbc56d4f73/source/common/router/upstream_request.cc#L770) could do either: 
1) when the downstream connection send buffer fills up, do not block upstream H1 stream till the Envoy gets response headers. or
2) stops the per_try_timeout when the downstream is above high watermark and resume the timeout after the watermark goes below low watermark.

We are inclined to the first approach, given this would bring it in line with how HTTP/2 works today. And the 2nd approach may cause slowness in detection of any upstream issue or the implementation will get very complicated to  "correctly" track the time the downstream connection was blocked and only apply the unblocked time to the timer.

**Comment by danzh2010** [comment]
cc @alyssawilk @mattklein123 @KBaichoo

**Comment by alyssawilk** [comment]
@mattklein123  @snowp I'm inclined to think (1) is the way to go, lmk if you disagree!

tl;dr problems in-house when H2 downstream connection gets blocked, and tells H1 upstream connections to stop reading.  Headers arrive from upstream but aren't read from the kernel because we fully stop reading, so all the timers for "has the backend responded" are not disabled.  I prefer "don't block until we've read the headers" over "try to adjust all relevant timers when downstream buffers are blocking upstream reads"

**Comment by KBaichoo** [comment]
Agreed, #2 is not the way to go.

I think #1 makes sense, but we've ran into problems with different protocols flow control behaving differently. Thoughts on changing this so that the H1 watermark would be softer similar to H2 allowing it to process a "window" worth of data?

I think having it be configurable could be useful as in deployments with untrusted upstream, an attacker could leverage flow control doesn't stop headers.

**Comment by alyssawilk** [comment]
This attack already exists for H2 and H3 so I don't think we're making things meaningfully worse by extending to H1. I'd be inclined to add a knob for prior behavior if anyone asks for it, WDYT?

**Comment by KBaichoo** [comment]
Fair enough. Runtime guard sgtm.

**Comment by danzh2010** [comment]
Thanks for the discussion! FWIW, this is a repro test for this issue: https://github.com/envoyproxy/envoy/compare/main...danzh2010:envoy:pertrytimeout?expand=1.

Meanwhile I'm working on implementing approach 1 with runtime guard.

**Comment by mattklein123** [comment]
Sorry late to this but I agree (1) makes sense. I will review the PR.
